### PR TITLE
Refactor Span<T> to ease implementation of JIT intrinsics

### PR DESCRIPTION
**Pull request by jkotas**
- Introduce internal ByReference<T> type for byref fields and change Span to use it
- Generalize handling of byref-like types in the type loader
- Make DangerousGetPinnableReference public while I was on it

### CODE REVIEW

**Comment by jkotas** [comment]
cc @russellhadley @ahsonkhan @KrzysztofCwalina @joshfree @mattwarren

File: src/mscorlib/src/System/Runtime/CompilerServices/Unsafe.cs [hunk 1]
```diff
@@ -41,6 +41,18 @@ public static int SizeOf<T>()
         }
 
         /// <summary>
+        /// Reinterprets the given location as a reference to a value of type<typeparamref name="T"/>.
+        /// </summary>
+        [NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref T AsRef<T>(void * source)
+        {
+            // The body of this function will be replaced by the EE with unsafe code that just returns sizeof !!T
```
Comment (mikedn) [inline src/mscorlib/src/System/Runtime/CompilerServices/Unsafe.cs]: > returns sizeof !!T

Wrong comment perhaps? I don't size what `sizeof` has to do with `AsRef`.

Comment (jkotas) [inline src/mscorlib/src/System/Runtime/CompilerServices/Unsafe.cs]: Copy&paste ... fixed. Thanks.

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT arm Cross Debug Build

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test

**Comment by gkhanna79** [comment]
LGTM

**Comment by russellhadley** [comment]
LGTM - I'll get started on the intrinsic side.

**Pull request merged**